deepseg.cc

```diff
@@ -76,6 +76,29 @@ cv::Mat convert_rgb_to_yuyv( cv::Mat input ) {
 	return yuyv;
 }
 
+cv::Mat alpha_blend(cv::Mat srca, cv::Mat srcb, cv::Mat mask) {
+	// alpha blend two (8UC3) source images using a mask (8UC1, 255=>srca, 0=>srcb), adapted from:
+	// https://www.learnopencv.com/alpha-blending-using-opencv-cpp-python/
+	assert(srca.rows==srcb.rows);
+	assert(srca.cols==srcb.cols);
+	cv::Mat out = cv::Mat::zeros(srca.size(), srca.type());
+	uint8_t *optr = (uint8_t*)out.data;
+	uint8_t *aptr = (uint8_t*)srca.data;
+	uint8_t *bptr = (uint8_t*)srcb.data;
+	uint8_t *mptr = (uint8_t*)mask.data;
+	int npix = srca.rows * srca.cols;
+	for (int pix=0; pix<npix; ++pix) {
+		// blending weights
+		float aw=(float)(*mptr)/255.0, bw=1.0-aw;
+		// blend each channel byte
+		*optr = (uint8_t)( (float)(*aptr)*aw + (float)(*bptr)*bw ); ++aptr; ++bptr; ++optr;
+		*optr = (uint8_t)( (float)(*aptr)*aw + (float)(*bptr)*bw ); ++aptr; ++bptr; ++optr;
+		*optr = (uint8_t)( (float)(*aptr)*aw + (float)(*bptr)*bw ); ++aptr; ++bptr; ++optr;
```

Comment:
Put increments inline with the read/write accesses. Use some bit-shift/multiply algorithm to avoid slow floating point arithmetic.

Reply:
Yeah, this could be tidier, I cut/paste... will fix.
Now that I'm not using a float mask (I am in my fork where this came from), I can use integer math too.